src/lexer.mll

```diff
@@ -271,8 +271,61 @@ let raw_markup_target =
 let language_tag_char =
   ['a'-'z' 'A'-'Z' '0'-'9' '_' '-' ]
 
+rule reference_paren_content input start depth buffer = parse 
+  | '('
+    {
+      Buffer.add_string buffer (Lexing.lexeme lexbuf) ;
+       reference_paren_content input start (depth + 1) buffer lexbuf }
+  | ')'
+    { 
+      Buffer.add_string buffer (Lexing.lexeme lexbuf) ;
+      if depth = 0 then
+        ()
+      else
+        ( reference_paren_content input start (depth  - 1) buffer lexbuf ) }
+  | eof 
+    { warning
```

Comment:
It would definitely be possible, however, `{!(})}` is allowed as of now, and not allowing means knowing stuff about ocaml identifiers that odoc-parser does not need to know.
But its doable ! We should probably check for both if we want to go down that road. 
It would also provide a better output for this case : `{! ( .*{} } ... `

Reply:
We don't expect unbalanced `{}` and `()` in references so I think it's fine to report a nice error for both `{!(})}` and `{! ( .*{} }` rather than accept the first but have a bad error for the second.

Reply:
Ok Ill do that then

Reply:
The double eof warning is still happening. I'd suggest calling `reference_content` (instead of iterating) from `reference_paren_content` (in the `)` case).